src/udev/memory_id/dmidecode.c

```diff
@@ -0,0 +1,768 @@
+                "DIMM",
+                "TSOP",
+                "Row Of Chips",
+                "RIMM",
+                "SODIMM",
+                "SRIMM",
+                "FB-DIMM",
+                "Die" /* 0x10 */
+        };
+
+        if (code >= 0x01 && code <= 0x10)
+                return form_factor[code - 0x01];
+        return out_of_spec;
+}
+
+static void dmi_memory_device_set(unsigned int slot_num, uint8_t code)
+{
+        if (code == 0xFF)
+                printf("MEMORY_DEVICE_%u_SET=%s\n", slot_num, "Unknown");
+        else if (code != 0)
+                printf("MEMORY_DEVICE_%u_SET=%u\n", slot_num, code);
+}
+
+static const char *dmi_memory_device_type(uint8_t code)
+{
+        /* 7.18.2 */
+        static const char *type[] = {
+                "Other", /* 0x01 */
+                "Unknown",
+                "DRAM",
+                "EDRAM",
+                "VRAM",
+                "SRAM",
+                "RAM",
+                "ROM",
+                "Flash",
+                "EEPROM",
+                "FEPROM",
+                "EPROM",
+                "CDRAM",
+                "3DRAM",
+                "SDRAM",
+                "SGRAM",
+                "RDRAM",
+                "DDR",
+                "DDR2",
+                "DDR2 FB-DIMM",
+                "Reserved",
+                "Reserved",
+                "Reserved",
+                "DDR3",
+                "FBD2",
+                "DDR4",
+                "LPDDR",
+                "LPDDR2",
+                "LPDDR3",
+                "LPDDR4",
+                "Logical non-volatile device",
+                "HBM",
+                "HBM2" /* 0x21 */
+        };
+
+        if (code >= 0x01 && code <= 0x21)
+                return type[code - 0x01];
+        return out_of_spec;
+}
+
+static void dmi_memory_device_type_detail(unsigned int slot_num, uint16_t code)
+{
+        /* 7.18.3 */
+        static const char *detail[] = {
+                "Other", /* 1 */
+                "Unknown",
+                "Fast-paged",
+                "Static Column",
+                "Pseudo-static",
+                "RAMBus",
+                "Synchronous",
+                "CMOS",
+                "EDO",
+                "Window DRAM",
+                "Cache DRAM",
+                "Non-Volatile",
+                "Registered (Buffered)",
+                "Unbuffered (Unregistered)",
+                "LRDIMM"  /* 15 */
+        };
+        char list[172];		/* Update length if you touch the array above */
+
+        if ((code & 0xFFFE) == 0)
+                printf("MEMORY_DEVICE_%u_TYPE_DETAIL=%s\n", slot_num, "None");
+        else
+        {
+                int i, off = 0;
+
+                list[0] = '\0';
+                for (i = 1; i <= 15; i++)
+                        if (code & (1 << i))
+                                off += sprintf(list + off, off ? " %s" : "%s",
+                                               detail[i - 1]);
```

Comment:
btw, why do you append everything in a buffer first and then output it with one printf()? you could just output things directly with printf() or fputs() or so. After all stdio is buffered io and internally maintains such buffers anyway. Would be a lot less fragile that way, given the manually calulcated array sizes won't be necessary anymore then.